Round total disk space up to nearest GB value in quota computation

szewai opened the pull request:
#### 68c9c237cbb1fd36a04c0ee4b24fb2016d9c9590
<pre>
Round total disk space up to nearest GB value in quota computation
<a href="https://bugs.webkit.org/show_bug.cgi?id=255016">https://bugs.webkit.org/show_bug.cgi?id=255016</a>
rdar://problem/107711233

Reviewed by Youenn Fablet.

We plan to calculate quota based on total disk space, and we currently use FileSystem::volumeCapacity to get total disk
space. However, the result returned by FileSystem::volumeCapacity is not necessarily a multiple of 1 GB and might be
slightly different for machines with the same hard disk space, depending on implementation of file systems. To address
this privacy concern, we now round up the value returned by volumeCapacity() when doing quota computation.

The patch also contains two drive-by fixes:
1. Introduce standardReportedQuota used for reported quota when an origin has a small usage and a big quota.
2. Replace defaultThirdPartyOriginQuota with defaultThirdPartyOriginQuotaRatio, since quota of third-party orgin should
never be greater than that of first-party origin.

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::quotaManager):
(WebKit::NetworkStorageManager::originStorageManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp:
(WebKit::OriginQuotaManager::create):
(WebKit::OriginQuotaManager::OriginQuotaManager):
(WebKit::OriginQuotaManager::reportedQuota const):
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
(WebKit::OriginQuotaManager::create):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::createQuotaManager):
(WebKit::OriginStorageManager::OriginStorageManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::perThirdPartyOriginStorageQuota const): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::perOriginStorageQuota const):

Canonical link: <a href="https://commits.webkit.org/262713@main">https://commits.webkit.org/262713@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/502327ace356f3cd8f8bcca9c7611529af40c8cd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2253 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2284 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2359 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3179 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2284 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2374 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2334 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2033 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2273 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/2012 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2023 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3037 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1 "21 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2004 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1908 "5 flakes 142 failures") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2083 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2021 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3202 "Passed tests") | 
| | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2050 "Passed tests") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1863 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2065 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2012 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/592 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1991 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2186 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
```diff
@@ -100,7 +100,7 @@ void NetworkSessionCreationParameters::encode(IPC::Encoder& encoder) const
 
     encoder << unifiedOriginStorageLevel;
-    encoder << perOriginStorageQuota << perThirdPartyOriginStorageQuota << originQuotaRatio << totalQuotaRatio << volumeCapacityOverride;
+    encoder << perOriginStorageQuota << originQuotaRatio << totalQuotaRatio << volumeCapacityOverride;
```
Removal of perThirdPartyOriginStorageQuota is not really described in change log or in title.
Let's clarify this there.

szewai replied:
> How about rounding the volume capacity up to the GB and then keep the computation the way it is since the issue seems to be variance in volumeCapacity?

This might also work; I will make the change and update test.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -382,7 +384,9 @@ OriginStorageManager& NetworkStorageManager::originStorageManager(const WebCore:
             IPC::Connection::send(connection, Messages::NetworkProcessProxy::IncreaseQuota(sessionID, origin, identifier, currentQuota, currentUsage, requestedIncrease), 0);
         };
-        uint64_t quota = m_defaultOriginQuota;
+        // Use double for mutlipication to preserve precision.
```
s/mutlipication/multiplication.
Also we should probably move standardReportedQuota. closer to where it is actually used.

szewai replied:
standardReportedQuota is in `NetworkStorageManager` because `OriginStorageManager` or `OriginQuotaManager` does not keep track of `WebCore::ClientOrigin`, and thus does not know policy for third-party origin.

Do you mean to move `defaultStandardReportedQuota` to `OriginQuotaManager`?

youennf reviewed (commented):
How about rounding the volume capacity up to the GB and then keep the computation the way it is since the issue seems to be variance in volumeCapacity?

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -367,10 +370,13 @@ NetworkQuotaManager& NetworkStorageManager::quotaManager()
+                volumeCapacity = m_volumeCapacityOverride;
             else if (auto capacity = FileSystem::volumeCapacity(m_path))
-                quota = m_totalQuotaRatio.value() * capacity.value();
+                volumeCapacity = WTF::roundUpToMultipleOf(defaultVolumeCapacityUnit, *capacity);
```
```suggestion
                volumeCapacity = WTF::roundUpToMultipleOf(defaultVolumeCapacityUnit, * capacity);
```

szewai replied:
hmm it seems uncommon to add a space between * and `std::optional<>` to get its contained value?

youennf replied:
Right, I misread it.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -367,10 +370,13 @@ NetworkQuotaManager& NetworkStorageManager::quotaManager()
-                quota = m_totalQuotaRatio.value() * capacity.value();
+                volumeCapacity = WTF::roundUpToMultipleOf(defaultVolumeCapacityUnit, *capacity);
+            if (volumeCapacity)
+                quota = m_totalQuotaRatio.value() * volumeCapacity.value();
```
Could use * instead of value()

szewai committed:
Round total disk space up to nearest GB value in quota computation
https://bugs.webkit.org/show_bug.cgi?id=255016
rdar://problem/107711233

Reviewed by Youenn Fablet.

We plan to calculate quota based on total disk space, and we currently use FileSystem::volumeCapacity to get total disk
space. However, the result returned by FileSystem::volumeCapacity is not necessarily a multiple of 1 GB and might be
slightly different for machines with the same hard disk space, depending on implementation of file systems. To address
this privacy concern, we now round up the value returned by volumeCapacity() when doing quota computation.

The patch also contains two drive-by fixes:
1. Introduce standardReportedQuota used for reported quota when an origin has a small usage and a big quota.
2. Replace defaultThirdPartyOriginQuota with defaultThirdPartyOriginQuotaRatio, since quota of third-party orgin should
never be greater than that of first-party origin.

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::quotaManager):
(WebKit::NetworkStorageManager::originStorageManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp:
(WebKit::OriginQuotaManager::create):
(WebKit::OriginQuotaManager::OriginQuotaManager):
(WebKit::OriginQuotaManager::reportedQuota const):
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
(WebKit::OriginQuotaManager::create):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::createQuotaManager):
(WebKit::OriginStorageManager::OriginStorageManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::perThirdPartyOriginStorageQuota const): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::perOriginStorageQuota const):

Canonical link: https://commits.webkit.org/262713@main
